File: Assets/Scripts/Neo/SideChannels/EpisodeSideChannel.cs
using System;
using System.Text;
using UnityEngine;
using Unity.MLAgents;
using Unity.MLAgents.SideChannels;

public class EpisodeSideChannel : SideChannel
{
    public NavOpsEnvController NavOpsEnvController;

    /**
     * https://github.com/Unity-Technologies/ml-agents/blob/main/docs/Custom-SideChannels.md
     */
    public EpisodeSideChannel()
    {
        ChannelId = new Guid("0a6d3d29-0130-475c-a98c-ae665a752cbc");
    }

    protected override void OnMessageReceived(IncomingMessage msg)
    {
        var receivedString = msg.ReadString();
        Debug.Log($"From Python: {receivedString}");
    }

    public void SendEpisodeDoneToPython(bool blueWins)
    {
        Debug.Log($"[EpisodeSideChannel] SendEpisodeDoneToPython: {blueWins}");
        using (var msgOut = new OutgoingMessage())
        {
            msgOut.WriteBoolean(blueWins);
            QueueMessageToSend(msgOut);
        }
    }
}

File: Assets/Scripts/Neo/SideChannels/episode_side_channel.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-
import uuid

import numpy as np
from mlagents_envs.environment import UnityEnvironment
from mlagents_envs.side_channel.side_channel import SideChannel, IncomingMessage


class EpisodeSideChannel(SideChannel):

    def __init__(self):
        super().__init__(uuid.UUID("0a6d3d29-0130-475c-a98c-ae665a752cbc"))

    def on_message_received(self, msg: IncomingMessage):
        print(msg.read_string())


if __name__ == "__main__":
    pass
